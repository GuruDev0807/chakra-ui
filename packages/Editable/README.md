# EditableText

EditableText is used for inline renaming of some text. It appears as normal UI
text but transforms into a text input field when the user clicks or focuses on
it.

## Installation

```sh
npm i @chakra-ui/editable
```
