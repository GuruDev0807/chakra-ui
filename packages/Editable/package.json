{
  "name": "@chakra-ui/editable",
  "version": "0.0.1",
  "description": "React component used to build inline-editable components",
  "keywords": [
    "chakra ui",
    "react",
    "inline",
    "editable",
    "inline editable",
    "ui",
    "editable text",
    "text field",
    "component"
  ],
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "homepage": "https://github.com/chakra-ui/chakra-ui#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "module": "dist/editable.esm.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/chakra-ui/chakra-ui.git"
  },
  "bugs": {
    "url": "https://github.com/chakra-ui/chakra-ui/issues"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "start": "tsdx watch",
    "build": "tsdx build",
    "test": "tsdx test --env=jsdom",
    "lint": "tsdx lint src",
    "version": "yarn build"
  },
  "devDependencies": {
    "@types/jest": "24.0.25",
    "@types/react": "16.9.17",
    "@types/react-dom": "16.9.4",
    "husky": "3.1.0",
    "react": "16.12.0",
    "react-dom": "16.12.0",
    "ts-jest": "24.0.2",
    "tsdx": "0.12.1",
    "tslib": "1.10.0",
    "typescript": "3.7.4"
  },
  "dependencies": {
    "@chakra-ui/hooks": "0.0.1",
    "@chakra-ui/utils": "0.0.1"
  },
  "peerDependencies": {
    "react": "16.x",
    "react-dom": "16.x"
  }
}
