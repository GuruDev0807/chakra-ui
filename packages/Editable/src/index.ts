export * from "./Editable";
export * from "./Editable.hook";
