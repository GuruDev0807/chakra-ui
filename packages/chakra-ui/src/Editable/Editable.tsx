import * as React from "react";
import { Box, BoxProps } from "@chakra-ui/layout";
import { Merge } from "@chakra-ui/utils";

interface EditableOptions {
  /**
   * The value of the Editable in both edit & preview mode
   */
  value?: string;
  /**
   * The initial value of the Editable in both edit & preview mode
   */
  defaultValue?: string;
  /**
   * If `true`, the Editable will be disabled.
   */
  isDisabled?: boolean;
  /**
   * If `true`, the Editable will start with edit mode by default.
   */
  startWithEditView?: boolean;
  /**
   * If `true`, the read only view, has a `tabIndex` set to `0` so it can recieve focus via the keyboard or click.
   */
  isPreviewFocusable?: boolean;
  /**
   * If `true`, it'll update the value onBlur and turn off the edit mode.
   */
  submitOnBlur?: boolean;
  /**
   * Callback invoked when user changes input.
   */
  onChange?: (newValue: string) => void;
  /**
   * Callback invoked when user cancels input with the `Esc` key.
   * It provides the last confirmed value as argument.
   */
  onCancel?: (previousValue: string) => void;
  /**
   * Callback invoked when user confirms value with `enter` key or by blurring input.
   */
  onSubmit?: (newValue: string) => void;
  /**
   * Callback invoked once the user enters edit mode.
   */
  onEdit?: () => void;
  /**
   * If `true`, the input's text will be highlighted on focus.
   */
  selectAllOnFocus?: boolean;
  /**
   * The placeholder text when the value is empty.
   */
  placeholder?: string;
  /**
   * The content of the EditableText
   * Ideally only `EditablePreview` and `EditableInput` should
   * be the children but you add other elements too
   */
  children: React.ReactNode;
}

interface EditableContextValue {
  inputRef?: React.Ref<HTMLInputElement>;
  isEditing?: boolean;
  isDisabled?: boolean;
  placeholder?: string;
  onRequestEdit?(): void;
  submitOnBlur?: boolean;
  isPreviewFocusable?: boolean;
  value?: string;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  onSubmit?(): void;
  onCancel?(): void;
  onFocus?(): void;
}

interface EditableInputRenderProps {
  ref?: EditableContextValue["inputRef"];
  onBlur?: React.FocusEventHandler<HTMLElement>;
  value?: string;
  placeholder?: string;
  onChange?: EditableContextValue["onChange"];
  onKeyDown?: EditableContextValue["onKeyDown"];
}

type EditableProps<P, T> = Merge<BoxProps<P, T>, EditableOptions>;

const EditableContext = React.createContext<EditableContextValue>({});

const Editable = React.forwardRef(function Editable<P, T>(
  {
    value: valueProp,
    defaultValue,
    isDisabled,
    onChange,
    startWithEditView,
    onCancel,
    onSubmit,
    selectAllOnFocus = true,
    submitOnBlur = true,
    isPreviewFocusable = true,
    placeholder = "Click to edit...",
    children,
    ...rest
  }: EditableProps<P, T>,
  ref: React.Ref<T>,
) {
  const [isEditing, setIsEditing] = React.useState<boolean>(
    Boolean(startWithEditView && !isDisabled),
  );
  const { current: isControlled } = React.useRef<boolean>(valueProp != null);

  const [value, setValue] = React.useState<string>(defaultValue || "");
  const _value = isControlled ? valueProp : value;
  const [previousValue, setPreviousValue] = React.useState<string>(
    _value as string,
  );

  const inputRef = React.useRef<HTMLInputElement>(null);

  const onRequestEdit = () => {
    if (!isDisabled) {
      setIsEditing(true);
    }
  };

  React.useEffect(() => {
    if (isEditing && inputRef.current) {
      inputRef.current.focus();
      selectAllOnFocus && inputRef.current.select();
    }
  }, [isEditing, selectAllOnFocus]);

  const handleCancel = () => {
    setIsEditing(false);
    setValue(previousValue);
    if (value !== previousValue) {
      onChange && onChange(previousValue);
    }
    onCancel && onCancel(previousValue);
  };

  const handleSubmit = () => {
    setIsEditing(false);
    setPreviousValue(value);
    onSubmit && onSubmit(value);
  };

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target;
    if (!isControlled) {
      setValue(value);
    }
    onChange && onChange(value);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const { key } = event;
    if (key === "Escape") {
      handleCancel();
      return;
    }

    if (key === "Enter") {
      handleSubmit();
    }
  };

  const handleFocus = () => {
    if (selectAllOnFocus && inputRef.current) {
      inputRef.current.select();
    }
  };

  const childContext: EditableContextValue = {
    inputRef,
    isEditing,
    isDisabled,
    placeholder,
    onRequestEdit,
    submitOnBlur,
    isPreviewFocusable,
    value: _value,
    onKeyDown: handleKeyDown,
    onChange: handleChange,
    onSubmit: handleSubmit,
    onCancel: handleCancel,
    onFocus: handleFocus,
  };

  return (
    <EditableContext.Provider value={childContext}>
      <Box ref={ref} {...rest}>
        {typeof children === "function"
          ? children({
              isEditing,
              onSubmit: handleSubmit,
              onCancel: handleCancel,
              onRequestEdit,
            })
          : children}
      </Box>
    </EditableContext.Provider>
  );
}) as <P, T>(
  props: EditableProps<P, T>,
) => React.ReactElement<EditableProps<P, T>>;

const sharedProps = {
  fontSize: "inherit",
  fontWeight: "inherit",
  textAlign: "inherit",
  bg: "transparent",
  transition: "all 0.2s",
  borderRadius: "md",
  px: "3px",
  mx: "-3px",
};

function useEditableState() {
  const {
    isEditing,
    onSubmit,
    onCancel,
    onRequestEdit,
    isDisabled,
  } = React.useContext(EditableContext);
  return {
    isEditing,
    onSubmit,
    onCancel,
    onEdit: onRequestEdit,
    isDisabled,
  };
}

const EditablePreview = React.forwardRef(function EditablePreview<P>(
  props: BoxProps<P>,
  ref: React.Ref<HTMLSpanElement>,
) {
  const {
    isEditing,
    isDisabled,
    value,
    onRequestEdit,
    placeholder,
    isPreviewFocusable,
  } = React.useContext(EditableContext);
  const hasValue = value != null && value !== "";

  const getTabIndex = () => {
    if ((!isEditing || !isDisabled) && isPreviewFocusable) {
      return 0;
    }

    return null;
  };

  const styleProps = {
    ...sharedProps,
    cursor: "text",
    display: "inline-block",
    opacity: !hasValue ? 0.6 : undefined,
  };

  if (isEditing) {
    return null;
  }

  return (
    <Box
      ref={ref}
      as="span"
      aria-disabled={isDisabled}
      tabIndex={getTabIndex()}
      onFocus={onRequestEdit}
      {...styleProps}
      {...props}
    >
      {hasValue ? value : placeholder}
    </Box>
  );
}) as React.FC<BoxProps>;

const EditableInput = React.forwardRef(function EditableInput<P>(
  props: BoxProps<P>,
  ref: React.Ref<HTMLInputElement>,
) {
  const {
    inputRef,
    isEditing,
    onChange,
    onKeyDown,
    value,
    onSubmit,
    submitOnBlur,
    placeholder,
    isDisabled,
  } = React.useContext(EditableContext);

  if (!isEditing) {
    return null;
  }

  const styleProps = {
    ...sharedProps,
    width: "full",
    _placeholder: {
      opacity: "0.6",
    },
  };

  const renderProps: EditableInputRenderProps = {
    ref: inputRef,
    onBlur: (event: React.FocusEvent<HTMLElement>) => {
      if (submitOnBlur && onSubmit) {
        onSubmit();
      }
      if (props.onBlur) {
        props.onBlur(event);
      }
    },
    value,
    placeholder,
    onChange,
    onKeyDown,
  };

  return typeof props.children === "function" ? (
    props.children(renderProps)
  ) : (
    <Box
      ref={ref}
      aria-disabled={isDisabled}
      disabled={isDisabled}
      as="input"
      outline="none"
      _focus={{ shadow: "outline" }}
      {...renderProps}
      {...styleProps}
      {...props}
    />
  );
}) as React.FC<BoxProps>;

export { EditablePreview, Editable, EditableInput, useEditableState };
