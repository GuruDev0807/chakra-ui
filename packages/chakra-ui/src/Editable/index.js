/** @jsx jsx */
import { jsx } from "@emotion/core";
import propTypes from "prop-types";
import { createContext, useContext, useEffect, useRef, useState } from "react";
import PseudoBox from "../PseudoBox";
import Box from "../Box";

const EditableContext = createContext();

const Editable = ({
  value: valueProp,
  defaultValue,
  isDisabled,
  onChange,
  isEditing: isEditingProp,
  onCancel,
  onSubmit,
  selectAllOnFocus = true,
  submitOnBlur,
  isPreviewFocusable,
  placeholder = "Click to edit...",
  children,
  ...rest
}) => {
  const [isEditing, setIsEditing] = useState(isEditingProp && !isDisabled);
  const { current: isControlled } = useRef(valueProp != null);

  const [value, setValue] = useState(defaultValue || "");
  const _value = isControlled ? valueProp : value;
  const [previousValue, setPreviousValue] = useState(_value);

  const inputRef = useRef(null);

  const onRequestEdit = event => {
    if (!isDisabled) {
      setIsEditing(true);
    }
  };

  useEffect(() => {
    if (isEditing && inputRef.current) {
      inputRef.current.focus();
      selectAllOnFocus && inputRef.current.select();
    }
  }, [isEditing, selectAllOnFocus]);

  const handleCancel = () => {
    setIsEditing(false);
    setValue(previousValue);
    if (value !== previousValue) {
      onChange && onChange(previousValue);
    }
    onCancel && onCancel(previousValue);
  };

  const handleSubmit = () => {
    setIsEditing(false);
    setPreviousValue(value);
    onSubmit && onSubmit(value);
  };

  const handleChange = event => {
    const { value } = event.target;
    if (!isControlled) {
      setValue(value);
    }
    onChange && onChange(value);
  };

  const handleKeyDown = event => {
    const { key } = event;
    if (key === "Escape") {
      handleCancel();
      return;
    }

    if (key === "Enter") {
      handleSubmit();
    }
  };

  const handleFocus = event => {
    if (selectAllOnFocus) {
      inputRef.current.select();
    }
  };

  const childContext = {
    inputRef,
    isEditing,
    isDisabled,
    placeholder,
    onRequestEdit,
    submitOnBlur,
    isPreviewFocusable,
    value: _value,
    onKeyDown: handleKeyDown,
    onChange: handleChange,
    onSubmit: handleSubmit,
    onCancel: handleCancel,
    onFocus: handleFocus,
  };

  return (
    <EditableContext.Provider value={childContext}>
      <Box {...rest}>
        {typeof children === "function"
          ? children({
              isEditing,
              onSubmit: handleSubmit,
              onCancel: handleCancel,
              onRequestEdit,
            })
          : children}
      </Box>
    </EditableContext.Provider>
  );
};

const sharedProps = {
  fontSize: "inherit",
  fontWeight: "inherit",
  textAlign: "inherit",
  bg: "transparent",
  transition: "all 0.2s",
  borderRadius: "md",
  px: "3px",
  mx: "-3px",
};

export const EditablePreview = props => {
  const {
    isEditing,
    isDisabled,
    value,
    onRequestEdit,
    placeholder,
    isPreviewFocusable,
  } = useContext(EditableContext);
  const hasValue = value != null && value !== "";

  const getTabIndex = () => {
    if ((!isEditing || !isDisabled) && isPreviewFocusable) {
      return 0;
    }

    return null;
  };

  const styleProps = {
    ...sharedProps,
    cursor: "text",
    display: "inline-block",
    opacity: !hasValue ? 0.6 : undefined,
  };

  if (isEditing) {
    return null;
  }

  return (
    <PseudoBox
      as="span"
      aria-disabled={isDisabled}
      tabIndex={getTabIndex()}
      onFocus={onRequestEdit}
      {...styleProps}
      {...props}
    >
      {hasValue ? value : placeholder}
    </PseudoBox>
  );
};

export const EditableInput = props => {
  const {
    inputRef,
    isEditing,
    onChange,
    onKeyDown,
    value,
    onSubmit,
    submitOnBlur,
    placeholder,
  } = useContext(EditableContext);

  const styleProps = {
    ...sharedProps,
    width: "full",
    _placeholder: {
      opacity: "0.6",
    },
  };

  if (!isEditing) {
    return null;
  }

  return (
    <PseudoBox
      as="input"
      ref={inputRef}
      onBlur={() => {
        submitOnBlur && onSubmit();
      }}
      value={value}
      placeholder={placeholder}
      onChange={onChange}
      onKeyDown={onKeyDown}
      {...styleProps}
      {...props}
    />
  );
};

Editable.propTypes = {
  /** Text value of the controlled input */
  value: propTypes.string,
  /** Default text value of uncontrolled input. */
  defaultValue: propTypes.string,
  /**
   * Whether the text can be edited.
   * @default false
   */
  isDisabled: propTypes.bool,
  /**
   * Whether the component should start with the edit mode active
   * If `true`, the input is shown by default.
   * @default false
   */
  isEditing: propTypes.bool,
  /** Callback invoked when user changes input in any way.  */
  onChange: propTypes.func,
  /** Callback invoked when user cancels input with the `Esc` key. Receives last confirmed value. */
  onCancel: propTypes.func,
  /** Callback invoked when user confirms value with `enter` key or by blurring input. */
  onSubmit: propTypes.func,
  /** Callback invoked after the user enters edit mode. */
  onEdit: propTypes.func,
  /**
   * If `true`, the input's text will be highlighted on focus.
   * @default false
   */
  selectAllOnFocus: propTypes.bool,
  /**
   * Placeholder text when the value is empty.
   * @default "Click to Edit"
   */
  placeholder: propTypes.string,
  /** The content of the Editable
   *  Ideally only `EditablePreview` and `EditableInput` should
   *  be the children (but you add other elements too)
   */
  children: propTypes.node,
};

export default Editable;
