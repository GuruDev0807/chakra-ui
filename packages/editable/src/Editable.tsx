import { chakra, PropsOf } from "@chakra-ui/system"
import {
  createContext,
  cx,
  isFunction,
  ReactNodeOrRenderProp,
  mergeRefs,
} from "@chakra-ui/utils"
import * as React from "react"
import { forwardRef, Ref } from "react"
import {
  useEditable,
  UseEditableProps,
  UseEditableReturn,
} from "./Editable.hook"

type EditableContext = Omit<UseEditableReturn, "htmlProps">

const [EditableProvider, useEditableContext] = createContext<EditableContext>()

const StyledEditable = chakra("div", { themeKey: "Editable.Root" })

type RenderProps = Pick<
  UseEditableReturn,
  "isEditing" | "onSubmit" | "onCancel" | "onEdit"
>

type Omitted = "onChange" | "value" | "children" | "defaultValue"

export type EditableProps = UseEditableProps &
  Omit<PropsOf<typeof StyledEditable>, Omitted> & {
    children?: ReactNodeOrRenderProp<RenderProps>
  }

/**
 * Editable
 *
 * The wrapper that provides context and logic for all editable
 * components. It renders a `div`
 */
export const Editable = forwardRef((props: EditableProps, ref: Ref<any>) => {
  const { htmlProps, ...context } = useEditable(props)

  const { isEditing, onSubmit, onCancel, onEdit } = context

  const _className = cx("chakra-editable", props.className)

  const children = isFunction(props.children)
    ? props.children({ isEditing, onSubmit, onCancel, onEdit })
    : props.children

  return (
    <EditableProvider value={context}>
      <StyledEditable ref={ref} {...htmlProps} className={_className}>
        {children}
      </StyledEditable>
    </EditableProvider>
  )
})

const StyledPreview = chakra("span", { themeKey: "Editable.Preview" })

export type EditablePreviewProps = PropsOf<typeof StyledPreview>

/**
 * EditablePreview
 *
 * The `span` used to display the final value, in the `preview` mode
 */

export const EditablePreview = forwardRef(
  (props: EditablePreviewProps, ref: Ref<any>) => {
    const { getPreviewProps } = useEditableContext()
    const previewProps = getPreviewProps(props)
    previewProps.ref = mergeRefs(ref, previewProps.ref)

    const _className = cx("chakra-editable__preview", props.className)

    return <StyledPreview {...previewProps} className={_className} />
  },
)

const StyledInput = chakra("input", { themeKey: "Editable.Input" })

export type EditableInputProps = PropsOf<typeof StyledInput>

/**
 * EditableInput
 *
 * The input used in the `edit` mode
 */
export const EditableInput = forwardRef(
  (props: EditableInputProps, ref: Ref<HTMLInputElement>) => {
    const { getInputProps } = useEditableContext()
    const inputProps = getInputProps({ ...props, ref })

    const _className = cx("chakra-editable__input", props.className)

    return <StyledInput {...inputProps} className={_className} />
  },
)

/**
 * React hook use to gain access to the editable state and actions.
 */
export function useEditableState() {
  const {
    isEditing,
    onSubmit,
    onCancel,
    onEdit,
    isDisabled,
  } = useEditableContext()

  return {
    isEditing,
    onSubmit,
    onCancel,
    onEdit,
    isDisabled,
  }
}

/**
 * React hook use to create controls for the editable component
 */
export function useEditableControls() {
  const {
    isEditing,
    getEditButtonProps,
    getCancelButtonProps,
    getSubmitButtonProps,
  } = useEditableContext()

  return {
    isEditing,
    getEditButtonProps,
    getCancelButtonProps,
    getSubmitButtonProps,
  }
}
