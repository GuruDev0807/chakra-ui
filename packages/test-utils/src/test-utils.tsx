import { ChakraProvider } from "@chakra-ui/core"
import "@testing-library/jest-dom/extend-expect"
import {
  render as RTLRender,
  RenderOptions,
  fireEvent,
  RenderResult,
  Queries,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)
expect.extend(toHaveNoViolations)

const AllProviders: React.FC = ({ children }) => (
  <ChakraProvider resetCSS>{children}</ChakraProvider>
)

export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult

export function render(ui: React.ReactElement, options?: RenderOptions) {
  return RTLRender(ui, { wrapper: AllProviders, ...options })
}

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
