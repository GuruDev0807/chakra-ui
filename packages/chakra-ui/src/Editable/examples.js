import { storiesOf } from "@storybook/react";
import React, { Fragment } from "react";
import Editable, { EditableInput, EditablePreview } from ".";
import Box from "../Box";
import Button from "../Button";
import ButtonGroup from "../ButtonGroup";
import Flex from "../Flex";

const stories = storiesOf("Editable", module);
stories.addDecorator(story => {
  return (
    <Box maxWidth="lg" mx="auto" mt={6} p={6}>
      {story()}
    </Box>
  );
});

const ControlButtons = ({ isEditing, onSubmit, onCancel, onRequestEdit }) => (
  <Flex>
    {!isEditing ? (
      <Button size="sm" onClick={onRequestEdit}>
        Edit
      </Button>
    ) : (
      <ButtonGroup size="sm">
        <Button color="green" onClick={onSubmit}>
          Save
        </Button>
        <Button onClick={onCancel}>Cancel</Button>
      </ButtonGroup>
    )}
  </Flex>
);

stories.add("Default", () => (
  <Editable textAlign="center" defaultValue="testing" fontSize="2xl">
    {props => (
      <Fragment>
        <Flex>
          <EditablePreview flex="1" wordBreak="words" />
          <EditableInput flex="1" />
        </Flex>
        <ControlButtons {...props} />
      </Fragment>
    )}
  </Editable>
));
