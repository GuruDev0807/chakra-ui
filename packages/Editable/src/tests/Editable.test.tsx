import * as React from "react"
import { render } from "@chakra-ui/test-utils"
import { Editable, EditableInput, EditablePreview } from "../Editable"

test("should match snapshot", () => {
  const utils = render(
    <Editable defaultValue="testing">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )

  expect(utils.asFragment()).toMatchSnapshot()
})

test("should render properly", () => {
  const utils = render(
    <Editable defaultValue="testing">
      <EditablePreview data-testid="preview" />
      <EditableInput data-testid="input" />
    </Editable>,
  )

  const preview = utils.getByTestId("preview")
  const input = utils.getByTestId("input")

  expect(input).toHaveAttribute("hidden")
  expect(preview).toHaveTextContent("testing")
})
