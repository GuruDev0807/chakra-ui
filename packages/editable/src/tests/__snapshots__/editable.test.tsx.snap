// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="chakra-editable"
  >
    <span
      class="chakra-editable__preview css-gsjxgq"
      data-testid="preview"
      tabindex="0"
    >
      testing
    </span>
    <input
      class="chakra-editable__input css-1qupasc"
      data-testid="input"
      hidden=""
      value="testing"
    />
  </div>
</DocumentFragment>
`;
